### DISCUSSION TITLE
RTR "batch_init_session" using Falcon.py

### DISCUSSION THREAD

**Discussion by ITjoyboy** [CrowdStrike/falconpy#676] [Q&A] [answered]
Hi,
I was trying to execute "run" command on multiple hosts at the same time using "batch_admin_command" but in order to that I first need to initiate a batch session using "batch_init_session" falcon.py script but when i am running the python script it showing the below error=
------------------------------------------
**code:** 
from falconpy import RealTimeResponse
falcon = RealTimeResponse(client_id="###############################",
                          client_secret="############################"
                          )
target_hosts = ["######################", "#######################"]
response = falcon.batch_init_sessions(host_ids=target_hosts,
                                      timeout_duration="10m"
                                      )
print(response)
------------------------------------------
**Output** :
{'status_code': 500, 'headers': {}, 'body': {'errors': [{'message': "HTTPSConnectionPool(host='api.crowdstrike.com', port=443): Max retries exceeded with url: /real-time-response/combined/batch-init-session/v1?timeout_duration=10m (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:997)')))"}], 'resources': []}}

I can see this is a SSL certificate error but I am no expert nether in python nor in API so if you guys just help me I would be really thankful.

**Comment by jshcodes** [comment, marked as answer]
Hi @ITjoyboy -

For environments where SSL verification cannot be performed, you can disable it by passing the `ssl_verify` keyword and setting it to False.

An example of this can be found here: https://falconpy.io/Usage/Environment-Configuration.html#ssl-verify 

Thank you for the question!

**Reply by ITjoyboy** [reply]
Hi @jshcodes thanks for your help mate the issue is resolved now.
And sorry for bothering you I do have one more question for you if you don't mind.
Actually I want to install application through RTR on multiple machine and I was able to do that using Swagger. But now there is another scenario I want to install application on multiple hosts same as before but this time I want to install the .exe on a Host group and the thing is that I want all the hosts in that group which are online as well as offline to get the installation done, for the host which are offline I want them to connect with the cloud and execute the command when they get online.
So is there any way to do that using falconpy or any other methods?

**Reply by jshcodes** [reply]
Yes there is! You need to pass the `queue_offline` keyword when you call `batch_init_sessions`.

We have a [discussion post on this topic](https://github.com/CrowdStrike/falconpy/discussions/415).

The sample this post references can be found [here](https://github.com/CrowdStrike/falconpy/blob/main/samples/rtr/queued_execute.py).

Let us know if you have any questions! 😄

**Reply by ITjoyboy** [reply]
Hi @jshcodes I have was reading the above posts before sending you the reply above but the thing is that I was not able to understand the discussion "https://github.com/CrowdStrike/falconpy/discussions/415" & what the "https://github.com/CrowdStrike/falconpy/blob/main/samples/rtr/queued_execute.py" script does and where I need to put the relevant details in the code. 
Also in the discussion https://github.com/CrowdStrike/falconpy/discussions/415
If you could just help me with that it will be a great help mate.

**Reply by jshcodes** [reply]
In your code above, you are making a call to `batch_init_sessions`. 

Add the keyword `queue_offline` to this statement and set it to _`True`_. Then try executing your command against the host group again.

**Example**

```python
from falconpy import RealTimeResponse
falcon = RealTimeResponse(client_id="###############################",
client_secret="############################"
)
target_hosts = ["######################", "#######################"]
response = falcon.batch_init_sessions(host_ids=target_hosts,
                                      timeout_duration="10m",
                                      queue_offline=True
                                      )
print(response)
```

**Reply by ITjoyboy** [reply]
Okay got it, 
But Since this session is only for 10min then how can I make it for a longer time period because the machines can be offline even for more than hours or days. So would I need to rerun this script after every 10min or is there any way to make the script run continously in the Background.

**Reply by jshcodes** [reply]
You shouldn't need to run the script again. Once the command is queued, the agent will wait up to 7 days to execute the command.

So the process from a high level would be:

-  Create a batch session with `queue_offline` enabled.
- Issue your command against the host group.
- Run a loop for a brief period of time to collect the responses from all the hosts that are online. (i.e. Give the commands time to execute and return a response back to you via [RTR_CheckCommandStatus](https://www.falconpy.io/Service-Collections/Real-Time-Response.html#rtr_checkcommandstatus) or [RTR_CheckAdminCommandStatus](https://www.falconpy.io/Service-Collections/Real-Time-Response-Admin.html#rtr_checkadmincommandstatus).)
- If any are still pending, log the IDs and Batch IDs somewhere (or however you wish to track this), then schedule a cron job (or some other process) to check again after an hour or so.  
    - At the end of the 7th day, have the scheduled job report back any AIDs that did not return to service and then delete itself.

**Reply by ITjoyboy** [reply]
Thanks josh for the valuable information.
Could you tell me if the script listed in this article "https://github.com/CrowdStrike/falconpy/blob/main/samples/rtr/queued_execute.py" can be used to do the same or I have modify the script to get the above work done for me

**Reply by jshcodes** [reply]
It's more of a discussion example, so you'll want to modify it.  It was coded to wait until the first host in the batch returns to service, and does not implement follow up scheduling.

**Reply by ITjoyboy** [reply]
Ok got it, I will try to modify and run it in case I need any help I will get back to you.
Thanks again

**Reply by ITjoyboy** [reply]
Hi josh, 
I was trying to read and understand what changes should be made on the script in order to get the desired results.
There was argument "-f" to find hostnames matching the string provided in it. But what if the group doesn't have similar pattern for the hostnames.
